File: lib/src/ability.dart
import 'dart:convert';

class Ability {
  const Ability({
    required this.id,
    required this.type,
    required this.name,
    required this.value,
  });

  factory Ability.fromJson(Map<String, dynamic> json) {
    return Ability(
      id: json['id'] as int,
      type: json['type'] as String,
      name: json['name'] as String,
      value: json['value'] is int ? json['value'] as int : 0,
    );
  }

  factory Ability.fromJsonString(String jsonString) {
    return Ability.fromJson(
      json.decode(jsonString) as Map<String, dynamic>,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'type': type,
      'name': name,
      'value': value,
    };
  }

  String toJsonString() {
    return json.encode(toJson());
  }

  final int id;
  final String type;
  final String name;
  final int? value;
}

File: lib/src/ability.js
// ability.js

class Ability {
  /**
   * @param {Object} params
   * @param {number} params.id
   * @param {string} params.type
   * @param {string} params.name
   * @param {number} [params.value]
   */
  constructor({ id, type, name, value = 0 }) {
    this.id = id;
    this.type = type;
    this.name = name;
    this.value = value;
  }

  /**
   * @param {Object} json
   * @returns {Ability}
   */
  static fromJson(json) {
    return new Ability({
      id: json.id,
      type: json.type,
      name: json.name,
      value: json.value ?? 0,
    });
  }

  /**
   * @param {string} jsonString
   * @returns {Ability}
   */
  static fromJsonString(jsonString) {
    const jsonObj = JSON.parse(jsonString);
    return Ability.fromJson(jsonObj);
  }

  /**
   * @returns {Object}
   */
  toJson() {
    return {
      id: this.id,
      type: this.type,
      name: this.name,
      value: this.value,
    };
  }

  /**
   * @returns {string}
   */
  toJsonString() {
    return JSON.stringify(this.toJson());
  }
}

module.exports = { Ability };